x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java

```diff
@@ -304,12 +322,19 @@ public void onFailure(Exception e) {
         }
     }
 
+    public void getRoleMappings(Set<String> names, ActionListener<List<ExpressionRoleMapping>> listener) {
+        innerGetRoleMappings(
+            names,
+            listener.delegateFailureAndWrap((l, mappings) -> l.onResponse(reservedRoleMappings.mergeWithReserved(mappings)))
```

Comment:
I think I am getting abit lost to where ClusterStateRoleMapper#resolveRoles for reservedRoleMappings is called so that is garunteed to have the mappings already?

Reply:
`ClusterStateRoleMapper::resolveRoles` is just wrapping `ClusterStateRoleMapper::getMappings` that looks at the current cluster state and resolves the roles from the current cluster state. `ClusterStateRoleMapper::resolveRoles` was previously called from the `CompositeRoleMapper` but with this change `getMappings` is instead called directly (so `resolveRoles` is skipped) as far as I can see.